runtime/win32.c

```diff
@@ -1258,3 +1257,34 @@ value caml_win32_xdg_defaults(void)
 
   CAMLreturn(result);
 }
+
+static INIT_ONCE get_temp_path_init_once = INIT_ONCE_STATIC_INIT;
+static BOOL CALLBACK get_temp_path_init_function(PINIT_ONCE InitOnce,
+                                                 PVOID Parameter,
+                                                 PVOID *lpContext)
+{
+  FARPROC pGetTempPath2W =
+    GetProcAddress(GetModuleHandle(L"KERNEL32.DLL"), "GetTempPath2W");
+  if (pGetTempPath2W)
+    *lpContext = pGetTempPath2W;
+  else
+    *lpContext = GetTempPath;
+  return TRUE;
+}
+
+value caml_win32_get_temp_path(void)
+{
+  CAMLparam0();
+  CAMLlocal1(path);
+  wchar_t buf[MAX_PATH+1];
+  DWORD (WINAPI *get_temp_path)(DWORD, LPWSTR);
+
+  InitOnceExecuteOnce(&get_temp_path_init_once, get_temp_path_init_function,
+                      NULL, (LPVOID *) &get_temp_path);
+
+  if (!get_temp_path(MAX_PATH+1, buf))
+    caml_win32_sys_error(GetLastError());
+  else
+    path = caml_copy_string_of_utf16(buf);
+  CAMLreturn(path);
```

Comment:
It's not equivalent to the implementation in `caml_win32_xdg_defaults`, though - it's allocating cons cells and both of its locals are used after a subsequent allocation. FWIW, you could just do:

```suggestion
  wchar_t buf[MAX_PATH+1];
  DWORD (WINAPI *get_temp_path)(DWORD, LPWSTR);

  InitOnceExecuteOnce(&get_temp_path_init_once, get_temp_path_init_function,
                      NULL, (LPVOID *) &get_temp_path);

  if (!get_temp_path(MAX_PATH+1, buf))
    caml_win32_sys_error(GetLastError());

  CAMLreturn(caml_copy_string_of_utf16(buf));
```

Reply:
I've added this too, thanks!